# Practical 1

In this practical we will set up a working Python3 development environment and will start familiarizing a bit with Python.

## Slides

Slides of the introduction can be found here: [Intro](docs/sciprolabintro.pdf)

## Setting up the environment

We will need to install several pieces of software to get a working programming environment suitable for this practical. In this section we will install everything that we will need in the next few weeks. 

Python3 is available for Windows, Linux and Mac.

### Our toolbox

Basically we will need to install some software to start off with. If you work on Windows or Mac, skip the following and go straight to the section "Windows/Mac installation"

#### Linux installation 

1. The Python interpreter. In this course we will use python version 3.x. A lot of information on python can be found on the [python web page](https://www.python.org/). Open a terminal and try typing in:

    ```python3```
    
    if you get an error like "python3 command not found" you need to install it, while if you get something like this:
    
    ![](img/console.png)
    
    you are already sorted, just type Ctrl-D to exit.

    Installation on a debian-like linux distribution (e.g. Ubuntu) can be done by typing the following commands on a terminal:

    ```sudo apt-get update```
    ```sudo apt-get install python3```



2. Install now the package manager pip, which is a very convenient tool to install python packages, with the following command:
    
    ```sudo apt-get install python3-pip```
    
    <div class="alert alert-warning">

    **Note:** 
    
    If pip is already installed in your system you will get a message like: 
    python3-pip is already the newest version (3.x.y)

    </div>
    
3. Finally, install the Integrated Development Environment (IDE) that we will be using. This is called Visual Studio Code and is available for all platforms. You can read about it [here](https://code.visualstudio.com/). Downloads for all platforms can be found [here](https://code.visualstudio.com/Download).

#### Windows/Mac installation 

1. The python interpreter. In this course we will use python version 3.x. A lot of information on python can be found on the [python web page](https://www.python.org/). Installers for Windows and Mac can be found at [this page](https://www.python.org/). Click on Download Python 3.6.x. When executing the installer, please remember to flag "Add Python 3.6.2 to PATH" and then click on Install now (see picture below). 

    ![](img/installnow.png)


2. Install now the Integrated Development Environment (IDE) that we will be using. This is called Visual Studio Code and is available for all platforms. You can read about it [here](https://code.visualstudio.com/). Downloads for all platforms can be found [here](https://code.visualstudio.com/Download).


## The console

To access the console on Linux just open a terminal and type:

```python3```

while in Windows you have to look for "Python" and run "Python 3.6".
The console should look like this:

   ![](img/pythonconsole.png)
   

Now we are all set to start interacting with the Python interpreter. Type in the following:

In [1]:
5 + 3

8

All as expected. The "In [1]" line is the input, while the "Out [1]" reports the output of the interpreter. Let's challenge python with some other operations:

In [2]:
12 / 5

2.4

In [3]:
1/133

0.007518796992481203

And some assignments:

In [4]:
a = 10
b = 7
s = a + b
d = a / b

print("sum is:",s, " division is:",d)

sum is: 17  division is: 1.4285714285714286


In the first four lines, values have been assigned to variables through the = operator. In the last line, the print function is used to display the output. For the time being, we will skip all the details and just notice that the print function somehow managed to get text and variables in input and coherently merged them in an output text. 
Although quite useful in some occasions, the console is quite limited therefore you can close it for now. To exit type Ctrl-D or exit().

## Visual Studio Code

Once you open the IDE Visual Studio Code you will see the welcome screen:

![](img/visual_studio_code.png)

You can find useful information on this tool [here](https://code.visualstudio.com/docs#vscode). Please spend some time having a look at that page.

Once you are done with it you can close this window. First thing to do is to set the python interpreter to use. Click on View --> Command Palette and type "Python" in the text search space. Select **Python: Select Workspace Interpreter** as shown in the picture below.

![](img/pythoninterpreter.png)

Finally, select the python version you want to use (e.g. Python3).

Now you can click on **Open Folder** to create a new folder to place all the scripts you are going to create. You can call it something like "EXERCISES". Next you can create a new file, *example1.py* (*.py* extension stands for python). 

Visual Studio Code will understand that you are writing Python code and will help you with valid syntax for your program.

<div class="alert alert-warning">

**Warning:** 

If you get the following error message:

![](img/pylint.png)

click on **Install Pylint** which is a useful tool to help your coding experience.
</div>


Add the following text to your *example1.py* file.

In [5]:
a = 10
b = 33
c = a / b

print("a:", a, " b:", b, " a/b=",c)

a: 10  b: 33  a/b= 0.30303030303030304


Please notice that Visual Studio Code will help you writing your Python scripts. For example, when you start writing the **print** line it will complete the code for you (**if the Pylint extension mentioned above is installed**), suggesting the functions that match the letters written. This useful feature is called *code completion* and, alongside suggesting possible matches, it also visualizes a description of the function and parameters it needs.

![](img/codecompletion.png)

Save the file (Ctrl+S as shortcut). By right-clicking in the code panel and selecting **Run Python File in Terminal** (see picture below) you can execute the code you have just written. 

![](img/pythonrun.png)

Upon clicking on *Run Python File in Terminal* a terminal panel should pop up in the lower section of the coding panel and the result shown above should be reported.

Saving script files like the *example1.py* above is also handy because they can be invoked several times (later on we will learn how to get inputs from the command line to make them more useful...). To do so, you just need to call the python intepreter passing the script file as parameter. From the folder containing the *example1.py* script:

```python3 example1.py```

will in fact return:

a: 10  b: 33  a/b= 0.30303030303030304


*Please take some time to familiarize with Visual Studio Code (creating files, saving files etc.) as in the next practicals we will take this ability for granted.*

## The debugger

## Exercises

1. Compute the area of a triangle having base 120 units and height 33. Assign the result to a variable named area and print it.

2. Compute the area of a square having side S equal to 145 units. Assign the result to a variable named area and print it.

3. If you have not done so already, put the two previous scripts in two separate files (e.g. triangle_area.py and square_area.py and execute them from the terminal).

4. Write a small script (trapezoid.py) that computes the area of a trapezoid having major base (MB) equal to 30 units, minor base (mb) equal to 12 and height (H) equal to 17. Print the resulting area. Try executing the script from inside Visual Studio Code and from the terminal.



